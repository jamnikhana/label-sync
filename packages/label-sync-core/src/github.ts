import * as assert from 'assert'
import Octokit from '@octokit/rest'

/**
 *
 * Github functions
 *
 */

export interface GithubRepository {
  id?: number
  owner: {
    id?: number
    login: string
  }
  name: string
  full_name: string
}

/**
 *
 * Converts Github repository name to GithubRepository
 *
 * @param name
 */
export function getRepositoryFromName(name: string): GithubRepository | null {
  try {
    const [owner, repo] = name.split('/')

    assert.ok(typeof owner === 'string')
    assert.ok(typeof repo === 'string')

    return {
      owner: { login: owner },
      name: repo,
      full_name: name,
    }
  } catch (err) {
    return null
  }
}

export interface GithubIssue {
  id: number
  number: number
  state: string
  title: string
  body: string
  labels: GithubLabel[]
}

/**
 *
 * Obtains all issues in a particular repository.
 *
 * @param github
 * @param repository
 */
export const getRepositoryIssues = paginate<GithubIssue>(
  (github, repository, page) =>
    github.issues.listForRepo({
      repo: repository.name,
      owner: repository.owner.login,
      page: page,
      per_page: 100,
    }),
  100,
)

// export interface GithubPullRequest {
//   id: number
//   number: number
//   base: { repo: GithubRepository }
//   state: string
//   title: string
//   body: string
//   labels: GithubLabel[]
// }

// /**
//  *
//  * Obtains all pull requests in a repository.
//  *
//  * @param github
//  * @param repository
//  */
// export const getRepositoryPullRequests = paginate<GithubPullRequest>(
//   (github, repository, page) =>
//     github.pulls.list({
//       repo: repository.name,
//       owner: repository.owner.login,
//       page: page,
//       per_page: 100,
//     }),
//   100,
// )

export interface GithubLabel {
  id?: number
  node_id?: string
  url?: string
  name: string
  description: string
  color: string
  default: boolean
}

/**
 *
 * Obtains all labels used in particular repository.
 *
 * @param github
 * @param repository
 */
export const getRepositoryLabels = paginate<GithubLabel>(
  (github, repository, page) =>
    github.issues.listLabelsForRepo({
      repo: repository.name,
      owner: repository.owner.login,
      page: page,
      per_page: 100,
    }),
  100,
)

/* Helper functions */

/**
 *
 * Paginates a Github function.
 *
 * @param fn
 * @param size
 */
function paginate<T>(
  fn: (
    github: Octokit,
    repository: GithubRepository,
    page: number,
  ) => Promise<Octokit.Response<T[]>>,
  size: number,
  page: number = 1,
): (github: Octokit, repository: GithubRepository) => Promise<T[]> {
  return (github: Octokit, repository: GithubRepository) =>
    fn(github, repository, page).then(async res => {
      /* Handle error case */
      if (res.status !== 200) {
        return Promise.reject()
      }

      /**
       * Return data if its size is less than treshold.
       * Paginate further otherwise.
       */
      if (res.data.length < size) {
        return res.data
      } else {
        return paginate(fn, size, page + 1)(github, repository).then(
          remainings => [...res.data, ...remainings],
        )
      }
    })
}
