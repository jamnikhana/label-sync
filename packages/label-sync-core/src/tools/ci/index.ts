export { Config, getRepositoriesFromConfiguration } from './config'
export { handleSync } from './sync'
