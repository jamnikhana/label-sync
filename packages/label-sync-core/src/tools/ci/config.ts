import { GithubRepository, getRepositoryFromName } from '../../github'
import { RepositoryConfig } from '../../types'
import { withDefault } from '../../utils'

export interface Config {
  [repository: string]: RepositoryConfig
}

export type ConfigError = { repository: string; message: string }

/**
 *
 * Hydrates repositories from configuration file.
 *
 * @param configuration
 */
export function getRepositoriesFromConfiguration(config: Config) {
  const repositories = Object.keys(config).reduce<{
    configurations: { repository: GithubRepository; config: RepositoryConfig }[]
    errors: ConfigError[]
  }>(
    (acc, repositoryName) => {
      const repository = getRepositoryFromName(repositoryName)

      if (!repository) {
        return {
          ...acc,
          errors: acc.errors.concat({
            repository: repositoryName,
            message: `Cannot decode the provided repository name ${repositoryName}`,
          }),
        }
      }

      return {
        ...acc,
        configurations: acc.configurations.concat({
          repository: repository,
          config: hydrateRepositoryConfig(config[repositoryName]),
        }),
      }
    },
    { configurations: [], errors: [] },
  )

  return repositories

  /**
   * Helpers
   */
  function hydrateRepositoryConfig(config: RepositoryConfig) {
    return {
      strict: withDefault(false)(config.strict),
      labels: config.labels,
    }
  }
}
