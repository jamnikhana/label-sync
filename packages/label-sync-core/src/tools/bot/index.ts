import mls from 'multilines'
import Octokit from '@octokit/rest'
import * as probot from 'probot'

import { Config } from '../ci'
import { GithubIssue, GithubRepository } from '../../github'
import { RepositoryManifest, getRepositoryManifest } from '../../manifest'
import { assignSiblingsToIssue } from '../../handlers/siblings/siblings'
import { withDefault } from '../../utils'
import { getRepositoriesFromConfiguration } from '../ci/config'

export interface BotOptions {
  logger: Console
  githubToken: string
}

/**
 *
 * Creates a Github Bot instance based on provided configuration.
 *
 * @param config
 * @param logger
 */
export async function getGithubBot(
  config: Config,
  options: BotOptions,
): Promise<((app: probot.Application) => void)> {
  /* Options */

  const logger = withDefault(console)(options.logger)

  /* Authentication */

  const client = new Octokit({
    headers: {
      accept: 'application/vnd.github.symmetra-preview+json',
    },
  })

  client.authenticate({
    type: 'app',
    token: options.githubToken,
  })

  /* Generate manifest */

  const manifest = await generateBotManifest(client, config)

  /* Github Bot */

  const bot = (app: probot.Application) => {
    app.on('issues.labeled', async (context: probot.Context) => {
      const repository: GithubRepository = context.payload.repository
      const issue: GithubIssue = context.payload.issue

      /* Check repository configuration */
      if (!Object.keys(manifest).includes(repository.full_name)) {
        logger.log(`No such repository configuration, ${repository}.`)
        return
      }

      const res = await assignSiblingsToIssue(
        context.github,
        repository,
        issue,
        manifest[repository.full_name],
      )

      context.log.info(
        mls`
          | Successfully applied siblings to issue ${issue.number}: 
          | ${res}
          `,
      )
    })

    app.on('pull_request.labeled', async (context: probot.Context) => {
      const repository: GithubRepository = context.payload.repository
      const pr: GithubIssue = context.payload.pull_request

      /* Check repository configuration */
      if (!Object.keys(manifest).includes(repository.full_name)) {
        context.log.info(`No such repository configuration, ${repository}.`)
        return
      }

      const res = await assignSiblingsToIssue(
        context.github,
        repository,
        pr,
        manifest[repository.full_name],
      )

      context.log.info(
        mls`
          | Successfully applied siblings to ${pr.number} pull request: 
          | ${res}
          `,
      )
    })
  }

  return bot

  /* Helper functions */

  /**
   *
   * Generates manifests for each repo and combines them into one.
   *
   * @param config
   */
  async function generateBotManifest(
    github: Octokit,
    config: Config,
  ): Promise<{ [repository: string]: RepositoryManifest }> {
    const repositories = getRepositoriesFromConfiguration(config)

    if (repositories.errors.length !== 0) {
      throw new Error(`Couldn't generate manifest!`)
    }

    const manifests = await Promise.all(
      repositories.configurations.map(async repository => {
        const manifest = await getRepositoryManifest(
          github,
          repository.repository,
          repository.config,
        )

        return {
          repository: repository,
          manifest: manifest,
        }
      }),
    )

    const manifest = manifests.reduce((acc, manifest) => {
      if (manifest.manifest.status === 'err') {
        throw new Error(
          `Couldn't generate manifest: ${manifest.manifest.message}.`,
        )
      }

      return {
        ...acc,
        [manifest.repository.repository.full_name]: manifest.manifest.manifest,
      }
    }, {})

    return manifest
  }
}
