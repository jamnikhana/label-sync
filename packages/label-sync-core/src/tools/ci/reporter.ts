import chalk from 'chalk'
import mls from 'multilines'

import { GithubRepository } from '../../github'
import {
  LabelSyncReport,
  createTerminalReport as createLabelSyncReport,
} from '../../handlers/labels'
import {
  SiblingSyncReport,
  createTerminalReport as createSiblingSyncReport,
} from '../../handlers/siblings'
import { RepositoryManifest } from '../../manifest'
import { RepositoryConfig } from '../../types'

import { Config, ConfigError } from './config'
import { SyncOptions } from './sync'

export type SyncReport = {
  config: Config
  options: SyncOptions
  syncs: RepositorySyncReport[]
  configErrors: ConfigError[]
}

export type RepositorySyncReport =
  | {
      status: 'success'
      repository: GithubRepository
      config: RepositoryConfig
      manifest: RepositoryManifest
      labels: LabelSyncReport
      siblings: SiblingSyncReport
    }
  | {
      status: 'error'
      message: string
      repository: GithubRepository
      config: RepositoryConfig
    }

/**
 *
 * Creates a human readable terminal report of CI Sync.
 * (Uses chalk to make report more lively.)
 *
 * @param report
 */
export function createTerminalReport(report: SyncReport): string {
  return mls`
    | ${chalk.bgBlueBright('Label Sync Report')}
    | ${chalk.gray('This is an autogenerated report for your project.')}
    | (dry run: "${report.options.dryRun}")
    | 
    |
    | Hey 😁 we ran some things. Check the changes we made:
    |
    | ${syncsList(report.syncs)}
    |
    | We had some difficulties with these projects:
    |
    | ${configErrorsList(report.configErrors)}
  `

  function syncsList(reports: RepositorySyncReport[]): string {
    return reports
      .map(report => {
        switch (report.status) {
          case 'success': {
            return mls`
              | ${createLabelSyncReport(report.labels)}
              |
              | ${createSiblingSyncReport(report.siblings)}
            `
          }

          case 'error': {
            return mls`
              | ${chalk.bgRed(report.repository.full_name)}
              | ${report.message}
            `
          }
        }
      })
      .join('\n')
  }

  function configErrorsList(reports: ConfigError[]): string {
    if (reports.length === 0) return 'No problems!'
    return reports.map(report => report.message).join('\n')
  }
}
