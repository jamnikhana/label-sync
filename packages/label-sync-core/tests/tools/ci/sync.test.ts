import { handleSync, Config } from '../../../src/tools/ci'

import * as fixtures from '../../__fixtures__/github'

describe('ci', () => {
  test('performs successful dryrun', async () => {
    const client = fixtures.githubClient()
    const config: Config = {
      'maticzav/label-sync': {
        labels: {
          basic: '123123',
        },
        strict: false,
      },
    }

    const res = await handleSync(client as any, config, {
      dryRun: true,
      skipSiblingSync: false,
    })

    expect(client.issues.createLabel).toBeCalledTimes(0)
    expect(client.issues.updateLabel).toBeCalledTimes(0)
    expect(client.issues.deleteLabel).toBeCalledTimes(0)
    expect(client.issues.addLabels).toBeCalledTimes(0)
    expect(res).toMatchSnapshot()
  })

  test('perform successful sync', async () => {
    const client = fixtures.githubClient()
    const config: Config = {
      'maticzav/label-sync': {
        labels: {
          basic: '123123',
        },
        strict: false,
      },
      'maticzav/graphql-shield': {
        labels: {
          basic: '234234',
          'kind/bug': {
            color: 'ff0000',
            siblings: ['bug/0-no-reproduction'],
          },
          'bug/0-no-reproduction': {
            color: '00ff00',
            description: 'No reproduction available.',
          },
        },
        strict: false,
      },
    }

    const res = await handleSync(client as any, config, {
      dryRun: false,
      skipSiblingSync: false,
    })

    expect(client.issues.createLabel).toBeCalledTimes(1)
    expect(client.issues.updateLabel).toBeCalledTimes(3)
    expect(client.issues.deleteLabel).toBeCalledTimes(0)
    expect(client.issues.addLabels).toBeCalledTimes(2)
    expect(res).toMatchSnapshot()
  })

  test('perform successful sync and skips siblings', async () => {
    const client = fixtures.githubClient()
    const config: Config = {
      'maticzav/label-sync': {
        labels: {
          basic: '123123',
        },
        strict: false,
      },
      'maticzav/graphql-shield': {
        labels: {
          basic: '234234',
          'kind/bug': {
            color: 'ff0000',
            siblings: ['bug/0-no-reproduction'],
          },
          'bug/0-no-reproduction': {
            color: '00ff00',
            description: 'No reproduction available.',
          },
        },
        strict: false,
      },
    }

    const res = await handleSync(client as any, config, {
      dryRun: false,
      skipSiblingSync: true,
    })

    expect(client.issues.createLabel).toBeCalledTimes(1)
    expect(client.issues.updateLabel).toBeCalledTimes(3)
    expect(client.issues.deleteLabel).toBeCalledTimes(0)
    expect(client.issues.addLabels).toBeCalledTimes(0)
    expect(res).toMatchSnapshot()
  })

  test('correctly reports configuration error', async () => {
    const client = fixtures.githubClient()
    const config: Config = {
      'label-sync': {
        labels: {
          basic: '123123',
          complex: {
            color: '321321',
            siblings: ['doesntexist'],
          },
        },
        strict: true,
      },
    }

    const res = await handleSync(client as any, config, {
      dryRun: true,
      skipSiblingSync: false,
    })

    expect(client.issues.createLabel).toBeCalledTimes(0)
    expect(client.issues.updateLabel).toBeCalledTimes(0)
    expect(client.issues.deleteLabel).toBeCalledTimes(0)
    expect(client.issues.addLabels).toBeCalledTimes(0)
    expect(res).toEqual({
      config: config,
      configErrors: [
        {
          message: 'Cannot decode the provided repository name label-sync',
          repository: 'label-sync',
        },
      ],
      options: { dryRun: true, skipSiblingSync: false },
      syncs: [],
    })
  })

  test('correctly reports repository configuration error', async () => {
    const client = fixtures.githubClient()
    const config: Config = {
      'maticzav/label-sync': {
        labels: {
          basic: '123123',
          complex: {
            color: '321321',
            siblings: ['doesntexist', 'kind/missing'],
          },
        },
        strict: true,
      },
    }

    const res = await handleSync(client as any, config, {
      dryRun: true,
      skipSiblingSync: false,
    })

    expect(client.issues.createLabel).toBeCalledTimes(0)
    expect(client.issues.updateLabel).toBeCalledTimes(0)
    expect(client.issues.deleteLabel).toBeCalledTimes(0)
    expect(client.issues.addLabels).toBeCalledTimes(0)
    expect(res).toEqual({
      config: {
        'maticzav/label-sync': {
          labels: {
            basic: '123123',
            complex: {
              color: '321321',
              siblings: ['doesntexist', 'kind/missing'],
            },
          },
          strict: true,
        },
      },
      configErrors: [],
      options: { dryRun: true, skipSiblingSync: false },
      syncs: [
        {
          config: {
            labels: {
              basic: '123123',
              complex: {
                color: '321321',
                siblings: ['doesntexist', 'kind/missing'],
              },
            },
            strict: true,
          },
          message:
            'Error generating manifest: Labels doesntexist, kind/missing are not defined',
          repository: {
            full_name: 'maticzav/label-sync',
            name: 'label-sync',
            owner: { login: 'maticzav' },
          },
          status: 'error',
        },
      ],
    })
  })
})
