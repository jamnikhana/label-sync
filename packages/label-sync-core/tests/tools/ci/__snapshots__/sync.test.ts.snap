// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ci perform successful sync 1`] = `
Object {
  "config": Object {
    "maticzav/graphql-shield": Object {
      "labels": Object {
        "basic": "234234",
        "bug/0-no-reproduction": Object {
          "color": "00ff00",
          "description": "No reproduction available.",
        },
        "kind/bug": Object {
          "color": "ff0000",
          "siblings": Array [
            "bug/0-no-reproduction",
          ],
        },
      },
      "strict": false,
    },
    "maticzav/label-sync": Object {
      "labels": Object {
        "basic": "123123",
      },
      "strict": false,
    },
  },
  "configErrors": Array [],
  "options": Object {
    "dryRun": false,
    "skipSiblingSync": false,
  },
  "syncs": Array [
    Object {
      "config": Object {
        "labels": Object {
          "basic": "123123",
        },
        "strict": false,
      },
      "labels": Object {
        "additions": Array [],
        "config": Object {
          "labels": Object {
            "basic": "123123",
          },
          "strict": false,
        },
        "options": Object {
          "dryRun": false,
          "skipSiblingSync": false,
        },
        "removals": Array [
          Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "bug/no-reproduction",
          },
          Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "kind/bug",
          },
        ],
        "repository": Object {
          "full_name": "maticzav/label-sync",
          "name": "label-sync",
          "owner": Object {
            "login": "maticzav",
          },
        },
        "updates": Array [
          Object {
            "color": "123123",
            "current_name": "basic",
            "description": "",
            "name": "basic",
            "owner": "maticzav",
            "repo": "label-sync",
          },
        ],
      },
      "manifest": Object {
        "basic": Object {
          "label": Object {
            "color": "f266f4",
            "default": false,
            "description": "",
            "name": "basic",
          },
          "siblings": Array [],
        },
        "bug/no-reproduction": Object {
          "label": Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "bug/no-reproduction",
          },
          "siblings": Array [],
        },
        "kind/bug": Object {
          "label": Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "kind/bug",
          },
          "siblings": Array [],
        },
      },
      "repository": Object {
        "full_name": "maticzav/label-sync",
        "name": "label-sync",
        "owner": Object {
          "login": "maticzav",
        },
      },
      "siblings": Object {
        "issues": Array [
          Object {
            "issue": Object {
              "body": "Very serious issue.",
              "id": 1421,
              "labels": Array [
                Object {
                  "color": "333333",
                  "default": false,
                  "description": "",
                  "name": "bug/no-reproduction",
                },
                Object {
                  "color": "f266f4",
                  "default": false,
                  "description": "",
                  "name": "basic",
                },
                Object {
                  "color": "333333",
                  "default": false,
                  "description": "",
                  "name": "kind/bug",
                },
              ],
              "number": 1,
              "state": "",
              "title": "Issue Title",
            },
            "siblings": Array [],
          },
          Object {
            "issue": Object {
              "body": "Not so serious issue.",
              "id": 1123,
              "labels": Array [
                Object {
                  "color": "333333",
                  "default": false,
                  "description": "",
                  "name": "bug/no-reproduction",
                },
                Object {
                  "color": "f266f4",
                  "default": false,
                  "description": "",
                  "name": "basic",
                },
                Object {
                  "color": "333333",
                  "default": false,
                  "description": "",
                  "name": "kind/bug",
                },
              ],
              "number": 5,
              "state": "",
              "title": "Another Issue Title",
            },
            "siblings": Array [],
          },
        ],
        "manifest": Object {
          "basic": Object {
            "label": Object {
              "color": "f266f4",
              "default": false,
              "description": "",
              "name": "basic",
            },
            "siblings": Array [],
          },
          "bug/no-reproduction": Object {
            "label": Object {
              "color": "333333",
              "default": false,
              "description": "",
              "name": "bug/no-reproduction",
            },
            "siblings": Array [],
          },
          "kind/bug": Object {
            "label": Object {
              "color": "333333",
              "default": false,
              "description": "",
              "name": "kind/bug",
            },
            "siblings": Array [],
          },
        },
        "options": Object {
          "dryRun": false,
          "skipSiblingSync": false,
        },
        "repository": Object {
          "full_name": "maticzav/label-sync",
          "name": "label-sync",
          "owner": Object {
            "login": "maticzav",
          },
        },
      },
      "status": "success",
    },
    Object {
      "config": Object {
        "labels": Object {
          "basic": "234234",
          "bug/0-no-reproduction": Object {
            "color": "00ff00",
            "description": "No reproduction available.",
          },
          "kind/bug": Object {
            "color": "ff0000",
            "siblings": Array [
              "bug/0-no-reproduction",
            ],
          },
        },
        "strict": false,
      },
      "labels": Object {
        "additions": Array [
          Object {
            "color": "00ff00",
            "description": "No reproduction available.",
            "name": "bug/0-no-reproduction",
            "owner": "maticzav",
            "repo": "graphql-shield",
          },
        ],
        "config": Object {
          "labels": Object {
            "basic": "234234",
            "bug/0-no-reproduction": Object {
              "color": "00ff00",
              "description": "No reproduction available.",
            },
            "kind/bug": Object {
              "color": "ff0000",
              "siblings": Array [
                "bug/0-no-reproduction",
              ],
            },
          },
          "strict": false,
        },
        "options": Object {
          "dryRun": false,
          "skipSiblingSync": false,
        },
        "removals": Array [
          Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "bug/no-reproduction",
          },
        ],
        "repository": Object {
          "full_name": "maticzav/graphql-shield",
          "name": "graphql-shield",
          "owner": Object {
            "login": "maticzav",
          },
        },
        "updates": Array [
          Object {
            "color": "234234",
            "current_name": "basic",
            "description": "",
            "name": "basic",
            "owner": "maticzav",
            "repo": "graphql-shield",
          },
          Object {
            "color": "ff0000",
            "current_name": "kind/bug",
            "description": "",
            "name": "kind/bug",
            "owner": "maticzav",
            "repo": "graphql-shield",
          },
        ],
      },
      "manifest": Object {
        "basic": Object {
          "label": Object {
            "color": "f266f4",
            "default": false,
            "description": "",
            "name": "basic",
          },
          "siblings": Array [],
        },
        "bug/0-no-reproduction": Object {
          "label": Object {
            "color": "00ff00",
            "default": false,
            "description": "No reproduction available.",
            "name": "bug/0-no-reproduction",
          },
          "siblings": Array [],
        },
        "bug/no-reproduction": Object {
          "label": Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "bug/no-reproduction",
          },
          "siblings": Array [],
        },
        "kind/bug": Object {
          "label": Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "kind/bug",
          },
          "siblings": Array [
            "bug/0-no-reproduction",
          ],
        },
      },
      "repository": Object {
        "full_name": "maticzav/graphql-shield",
        "name": "graphql-shield",
        "owner": Object {
          "login": "maticzav",
        },
      },
      "siblings": Object {
        "issues": Array [
          Object {
            "issue": Object {
              "body": "Very serious issue.",
              "id": 1421,
              "labels": Array [
                Object {
                  "color": "333333",
                  "default": false,
                  "description": "",
                  "name": "bug/no-reproduction",
                },
                Object {
                  "color": "f266f4",
                  "default": false,
                  "description": "",
                  "name": "basic",
                },
                Object {
                  "color": "333333",
                  "default": false,
                  "description": "",
                  "name": "kind/bug",
                },
              ],
              "number": 1,
              "state": "",
              "title": "Issue Title",
            },
            "siblings": Array [
              Object {
                "color": "00ff00",
                "default": false,
                "description": "No reproduction available.",
                "name": "bug/0-no-reproduction",
              },
            ],
          },
          Object {
            "issue": Object {
              "body": "Not so serious issue.",
              "id": 1123,
              "labels": Array [
                Object {
                  "color": "333333",
                  "default": false,
                  "description": "",
                  "name": "bug/no-reproduction",
                },
                Object {
                  "color": "f266f4",
                  "default": false,
                  "description": "",
                  "name": "basic",
                },
                Object {
                  "color": "333333",
                  "default": false,
                  "description": "",
                  "name": "kind/bug",
                },
              ],
              "number": 5,
              "state": "",
              "title": "Another Issue Title",
            },
            "siblings": Array [
              Object {
                "color": "00ff00",
                "default": false,
                "description": "No reproduction available.",
                "name": "bug/0-no-reproduction",
              },
            ],
          },
        ],
        "manifest": Object {
          "basic": Object {
            "label": Object {
              "color": "f266f4",
              "default": false,
              "description": "",
              "name": "basic",
            },
            "siblings": Array [],
          },
          "bug/0-no-reproduction": Object {
            "label": Object {
              "color": "00ff00",
              "default": false,
              "description": "No reproduction available.",
              "name": "bug/0-no-reproduction",
            },
            "siblings": Array [],
          },
          "bug/no-reproduction": Object {
            "label": Object {
              "color": "333333",
              "default": false,
              "description": "",
              "name": "bug/no-reproduction",
            },
            "siblings": Array [],
          },
          "kind/bug": Object {
            "label": Object {
              "color": "333333",
              "default": false,
              "description": "",
              "name": "kind/bug",
            },
            "siblings": Array [
              "bug/0-no-reproduction",
            ],
          },
        },
        "options": Object {
          "dryRun": false,
          "skipSiblingSync": false,
        },
        "repository": Object {
          "full_name": "maticzav/graphql-shield",
          "name": "graphql-shield",
          "owner": Object {
            "login": "maticzav",
          },
        },
      },
      "status": "success",
    },
  ],
}
`;

exports[`ci perform successful sync and skips siblings 1`] = `
Object {
  "config": Object {
    "maticzav/graphql-shield": Object {
      "labels": Object {
        "basic": "234234",
        "bug/0-no-reproduction": Object {
          "color": "00ff00",
          "description": "No reproduction available.",
        },
        "kind/bug": Object {
          "color": "ff0000",
          "siblings": Array [
            "bug/0-no-reproduction",
          ],
        },
      },
      "strict": false,
    },
    "maticzav/label-sync": Object {
      "labels": Object {
        "basic": "123123",
      },
      "strict": false,
    },
  },
  "configErrors": Array [],
  "options": Object {
    "dryRun": false,
    "skipSiblingSync": true,
  },
  "syncs": Array [
    Object {
      "config": Object {
        "labels": Object {
          "basic": "123123",
        },
        "strict": false,
      },
      "labels": Object {
        "additions": Array [],
        "config": Object {
          "labels": Object {
            "basic": "123123",
          },
          "strict": false,
        },
        "options": Object {
          "dryRun": false,
          "skipSiblingSync": true,
        },
        "removals": Array [
          Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "bug/no-reproduction",
          },
          Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "kind/bug",
          },
        ],
        "repository": Object {
          "full_name": "maticzav/label-sync",
          "name": "label-sync",
          "owner": Object {
            "login": "maticzav",
          },
        },
        "updates": Array [
          Object {
            "color": "123123",
            "current_name": "basic",
            "description": "",
            "name": "basic",
            "owner": "maticzav",
            "repo": "label-sync",
          },
        ],
      },
      "manifest": Object {
        "basic": Object {
          "label": Object {
            "color": "f266f4",
            "default": false,
            "description": "",
            "name": "basic",
          },
          "siblings": Array [],
        },
        "bug/no-reproduction": Object {
          "label": Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "bug/no-reproduction",
          },
          "siblings": Array [],
        },
        "kind/bug": Object {
          "label": Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "kind/bug",
          },
          "siblings": Array [],
        },
      },
      "repository": Object {
        "full_name": "maticzav/label-sync",
        "name": "label-sync",
        "owner": Object {
          "login": "maticzav",
        },
      },
      "siblings": null,
      "status": "success",
    },
    Object {
      "config": Object {
        "labels": Object {
          "basic": "234234",
          "bug/0-no-reproduction": Object {
            "color": "00ff00",
            "description": "No reproduction available.",
          },
          "kind/bug": Object {
            "color": "ff0000",
            "siblings": Array [
              "bug/0-no-reproduction",
            ],
          },
        },
        "strict": false,
      },
      "labels": Object {
        "additions": Array [
          Object {
            "color": "00ff00",
            "description": "No reproduction available.",
            "name": "bug/0-no-reproduction",
            "owner": "maticzav",
            "repo": "graphql-shield",
          },
        ],
        "config": Object {
          "labels": Object {
            "basic": "234234",
            "bug/0-no-reproduction": Object {
              "color": "00ff00",
              "description": "No reproduction available.",
            },
            "kind/bug": Object {
              "color": "ff0000",
              "siblings": Array [
                "bug/0-no-reproduction",
              ],
            },
          },
          "strict": false,
        },
        "options": Object {
          "dryRun": false,
          "skipSiblingSync": true,
        },
        "removals": Array [
          Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "bug/no-reproduction",
          },
        ],
        "repository": Object {
          "full_name": "maticzav/graphql-shield",
          "name": "graphql-shield",
          "owner": Object {
            "login": "maticzav",
          },
        },
        "updates": Array [
          Object {
            "color": "234234",
            "current_name": "basic",
            "description": "",
            "name": "basic",
            "owner": "maticzav",
            "repo": "graphql-shield",
          },
          Object {
            "color": "ff0000",
            "current_name": "kind/bug",
            "description": "",
            "name": "kind/bug",
            "owner": "maticzav",
            "repo": "graphql-shield",
          },
        ],
      },
      "manifest": Object {
        "basic": Object {
          "label": Object {
            "color": "f266f4",
            "default": false,
            "description": "",
            "name": "basic",
          },
          "siblings": Array [],
        },
        "bug/0-no-reproduction": Object {
          "label": Object {
            "color": "00ff00",
            "default": false,
            "description": "No reproduction available.",
            "name": "bug/0-no-reproduction",
          },
          "siblings": Array [],
        },
        "bug/no-reproduction": Object {
          "label": Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "bug/no-reproduction",
          },
          "siblings": Array [],
        },
        "kind/bug": Object {
          "label": Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "kind/bug",
          },
          "siblings": Array [
            "bug/0-no-reproduction",
          ],
        },
      },
      "repository": Object {
        "full_name": "maticzav/graphql-shield",
        "name": "graphql-shield",
        "owner": Object {
          "login": "maticzav",
        },
      },
      "siblings": null,
      "status": "success",
    },
  ],
}
`;

exports[`ci performs successful dryrun 1`] = `
Object {
  "config": Object {
    "maticzav/label-sync": Object {
      "labels": Object {
        "basic": "123123",
      },
      "strict": false,
    },
  },
  "configErrors": Array [],
  "options": Object {
    "dryRun": true,
    "skipSiblingSync": false,
  },
  "syncs": Array [
    Object {
      "config": Object {
        "labels": Object {
          "basic": "123123",
        },
        "strict": false,
      },
      "labels": Object {
        "additions": Array [],
        "config": Object {
          "labels": Object {
            "basic": "123123",
          },
          "strict": false,
        },
        "options": Object {
          "dryRun": true,
          "skipSiblingSync": false,
        },
        "removals": Array [
          Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "bug/no-reproduction",
          },
          Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "kind/bug",
          },
        ],
        "repository": Object {
          "full_name": "maticzav/label-sync",
          "name": "label-sync",
          "owner": Object {
            "login": "maticzav",
          },
        },
        "updates": Array [
          Object {
            "color": "123123",
            "default": false,
            "description": "",
            "name": "basic",
          },
        ],
      },
      "manifest": Object {
        "basic": Object {
          "label": Object {
            "color": "f266f4",
            "default": false,
            "description": "",
            "name": "basic",
          },
          "siblings": Array [],
        },
        "bug/no-reproduction": Object {
          "label": Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "bug/no-reproduction",
          },
          "siblings": Array [],
        },
        "kind/bug": Object {
          "label": Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "kind/bug",
          },
          "siblings": Array [],
        },
      },
      "repository": Object {
        "full_name": "maticzav/label-sync",
        "name": "label-sync",
        "owner": Object {
          "login": "maticzav",
        },
      },
      "siblings": Object {
        "issues": Array [
          Object {
            "issue": Object {
              "body": "Very serious issue.",
              "id": 1421,
              "labels": Array [
                Object {
                  "color": "333333",
                  "default": false,
                  "description": "",
                  "name": "bug/no-reproduction",
                },
                Object {
                  "color": "f266f4",
                  "default": false,
                  "description": "",
                  "name": "basic",
                },
                Object {
                  "color": "333333",
                  "default": false,
                  "description": "",
                  "name": "kind/bug",
                },
              ],
              "number": 1,
              "state": "",
              "title": "Issue Title",
            },
            "siblings": Array [],
          },
          Object {
            "issue": Object {
              "body": "Not so serious issue.",
              "id": 1123,
              "labels": Array [
                Object {
                  "color": "333333",
                  "default": false,
                  "description": "",
                  "name": "bug/no-reproduction",
                },
                Object {
                  "color": "f266f4",
                  "default": false,
                  "description": "",
                  "name": "basic",
                },
                Object {
                  "color": "333333",
                  "default": false,
                  "description": "",
                  "name": "kind/bug",
                },
              ],
              "number": 5,
              "state": "",
              "title": "Another Issue Title",
            },
            "siblings": Array [],
          },
        ],
        "manifest": Object {
          "basic": Object {
            "label": Object {
              "color": "f266f4",
              "default": false,
              "description": "",
              "name": "basic",
            },
            "siblings": Array [],
          },
          "bug/no-reproduction": Object {
            "label": Object {
              "color": "333333",
              "default": false,
              "description": "",
              "name": "bug/no-reproduction",
            },
            "siblings": Array [],
          },
          "kind/bug": Object {
            "label": Object {
              "color": "333333",
              "default": false,
              "description": "",
              "name": "kind/bug",
            },
            "siblings": Array [],
          },
        },
        "options": Object {
          "dryRun": true,
          "skipSiblingSync": false,
        },
        "repository": Object {
          "full_name": "maticzav/label-sync",
          "name": "label-sync",
          "owner": Object {
            "login": "maticzav",
          },
        },
      },
      "status": "success",
    },
  ],
}
`;
